src/v/compression/async_stream_zstd.cc

```diff
@@ -0,0 +1,262 @@
+  : _decompress_ctx(ctx._decompress_ctx)
+  , _d_workspace(std::move(ctx._d_workspace))
+  , _d_buffer(std::move(ctx._d_buffer)) {
+    ctx._decompress_ctx = nullptr;
+}
+
+async_stream_zstd::static_dctx&
+async_stream_zstd::static_dctx::operator=(static_dctx&& x) noexcept {
+    if (this != &x) {
+        _decompress_ctx = x._decompress_ctx;
+        _d_workspace = std::move(x._d_workspace);
+        _d_buffer = std::move(x._d_buffer);
+
+        x._decompress_ctx = nullptr;
+    }
+
+    return *this;
+}
+
+async_stream_zstd::async_stream_zstd(
+  size_t decompression_size, int number_of_contexts) {
+    _decompression_size = decompression_size;
+
+    for (auto i = 0; i < number_of_contexts; i++) {
+        _compression_ctx_pool.release_object(
+          static_cctx(default_compression_level));
+        _decompression_ctx_pool.release_object(static_dctx(decompression_size));
+    }
+}
+
+size_t async_stream_zstd::decompression_size() const {
+    return _decompression_size;
+}
+
+ss::future<iobuf> async_stream_zstd::compress(iobuf i_buf) {
+    ss::abort_source as;
+    auto ctx_scope = co_await _compression_ctx_pool.allocate_scoped_object(as);
+    auto& ctx = *ctx_scope;
+
+    // reset compressor ctx
+    ZSTD_CCtx_reset(ctx._compress_ctx, ZSTD_reset_session_only);
+
+    // NOTE: always enable content size. **decompression** depends on this
+    throw_if_error(
+      ZSTD_CCtx_setPledgedSrcSize(ctx._compress_ctx, i_buf.size_bytes()));
+
+    iobuf ret_buf;
+
+    // if the buffer is empty encode and return an empty zstd frame.
+    if (i_buf.empty()) {
+        size_t remaining = 0;
+
+        do {
+            ZSTD_outBuffer out = {
+              .dst = ctx._c_buffer.get_write(),
+              .size = ctx._c_buffer.size(),
+              .pos = 0};
+            remaining = ZSTD_endStream(ctx._compress_ctx, &out);
+            throw_if_error(remaining);
+            ret_buf.append(ctx._c_buffer.get(), out.pos);
+        } while (remaining > 0);
+
+        co_return std::move(ret_buf);
+    }
+
+    // otherwise compress and return the buffer.
+    for (auto it = i_buf.cbegin(); it != i_buf.cend();) {
+        ZSTD_inBuffer in = {.src = it->get(), .size = it->size(), .pos = 0};
+        ++it;
+
+        const auto mode = (it == i_buf.cend()) ? ZSTD_e_end : ZSTD_e_continue;
+
+        int finished;
+        do {
+            ZSTD_outBuffer out = {
+              .dst = ctx._c_buffer.get_write(),
+              .size = ctx._c_buffer.size(),
+              .pos = 0};
+
+            const auto remaining = ZSTD_compressStream2(
+              ctx._compress_ctx, &out, &in, mode);
+            throw_if_error(remaining);
+
+            ret_buf.append(ctx._c_buffer.get(), out.pos);
+            finished = (mode == ZSTD_e_end) ? (remaining == 0)
+                                            : (in.pos == in.size);
+
+            co_await ss::coroutine::maybe_yield();
+        } while (!finished);
+    }
+
+    co_return std::move(ret_buf);
+}
```

Comment:
AFAICT this is a pessimizing std::move, since move is implicit and inhibit [N]RVO which is "even better". clang-tidy probably complains.

Reply:
Removing the explicit move to allow for elision.